### PR TITLE
Sequential Supervisor

### PR DESCRIPTION
**Pull request by josef-widder**
Specification of a sequential supervisor that explains how 
- initialization
- verification
- attack detection

work together and which addresses several issues:

- https://github.com/informalsystems/tendermint-rs/issues/499
- https://github.com/informalsystems/tendermint-rs/pull/509
- https://github.com/tendermint/spec/issues/131
- https://github.com/informalsystems/tendermint-rs/issues/461

It uses new functions for the lightstore which are defined in #171 .

### CODE REVIEW

File: rust-spec/lightclient/supervisor/supervisor_001_draft.md [hunk 1]
```diff
@@ -0,0 +1,641 @@
...
+>
+> - do verification for the next height *h' != h*
+> - do cross-checking for height *h*. If we find an attack, we remove
+>   *h* from the lightstore.
+>
+> Thus, this concurrency model changes the semantics of the
+> lightstore (not all lightblocks are trusted; they may be removed if
+> we find a problem). Whether this is desirable, and whether the gain in
+> performance is worth it, we keep for future versions/discussion of
+> lightclient protocols.
```
Comment (josef-widder) [inline rust-spec/lightclient/supervisor/supervisor_001_draft.md:200]: @cezarad says: this text is outdated, please update

File: rust-spec/lightclient/supervisor/supervisor_001_draft.md [hunk 2]
```diff
@@ -0,0 +1,641 @@
...
+#### Main verification and detection logic
+
+#### **[LC-FUNC-MAIN-VERIF-DETECT.1]:**
+
+```go
+func VerifyAndDetect (lightStore LightStore, targetHeight Height)
+                     (LightStore, Result) {
+
+    b1, r1 = lightStore.Get(targetHeight)
+    if r1 = true &&
```
Comment (josef-widder) [inline rust-spec/lightclient/supervisor/supervisor_001_draft.md:556]:
```suggestion
    if r1 = true {
```

File: rust-spec/lightclient/supervisor/supervisor_001_draft.md [hunk 3]
```diff
@@ -0,0 +1,640 @@
...
+#### Main verification and detection logic
+
+#### **[LC-FUNC-MAIN-VERIF-DETECT.1]:**
+
+```go
+func VerifyAndDetect (lightStore LightStore, targetHeight Height)
+                     (LightStore, Result) {
+
+    b1, r1 = lightStore.Get(targetHeight)
+    if r1 == true {
```
Comment (josef-widder) [inline rust-spec/lightclient/supervisor/supervisor_001_draft.md:556]: @cezarad says fix identation

File: rust-spec/lightclient/supervisor/supervisor_001_draft.md [hunk 4]
```diff
@@ -0,0 +1,640 @@
...
+func Sequential-Supervisor (initdata LCInitData) (Error) {
+
+    lightStore,result := InitLightClient(initData);
+    if result != OK {
+        return result;
+    }
+
+    loop {
+        // get the next height
+        nextHeight := input();
```
Comment (josef-widder) [inline rust-spec/lightclient/supervisor/supervisor_001_draft.md:430]: @cezarad says: in the Rust implementation there are mutliple input actions : VerifytoLatest, LatestTrusted, and GetStatus. The information can be easily obtained from the lightstore. But add a comment for completeness

File: rust-spec/lightclient/supervisor/supervisor_001_draft.md [hunk 5]
```diff
@@ -0,0 +1,640 @@
...
+no light client attack up to height *h*, then the lightclient
+eventually puts the lightblock of height *h* in the lightstore and
+wait for another input.
+- otherwise, that is, if there
+is a light client attack on height *h*, then the light client
+must perform one of the following:
+    - it terminates on attack.
+    - it eventually puts the lightblock of height *h* in the lightstore and
+wait for another input.
+
```
Comment (josef-widder) [inline rust-spec/lightclient/supervisor/supervisor_001_draft.md:164]:
```suggestion

> Observe that the "existence of a lightclient" just means that some node has generated a conflicting block. It does not necessarily mean that a (faulty) peer sends such a block to "our" lightclient. Thus, even if there is an attack somewhere in the system, our lightclient might still continue to operate normally.

```

**Comment by josef-widder** [comment]
@cezarad, I did the changes you requested. I guess we can we can merge this draft.

**Review by cezarad — APPROVED** [review]

**Pull request merged**